### DISCUSSION TITLE
How can I use <bits/stdc++.h> on CP Editor

### DISCUSSION THREAD

**Discussion by Arius1408** [cpeditor/cpeditor#778] [Q&A] [answered]
Hi guys, I started to do my work on macOS lately, and I found it annoying to use Xcode for CP stuff. So I have GCC installed via HomeBrew and replaced the default g++ with g++-10 (kinda like [this](https://www.youtube.com/watch?v=0z-fCNNqfEg)).
So now I can execute C++ file with <bits/stdc++.h> on command line  like : `g++ main.cpp -o main && ./main` without problems. But when it come to CP Editor, I can't execute the same file although the execute command on my CP Editor is still `g++ -Wall`

My CP Editor messages : 

`[10:32:07] [Compiler] [Failed to open [/Volumes/External-Drive/Documents/C++ Projects/Proj3/script.cpp]. Do I have write permission?]
[10:32:07] [Compiler] [Compilation has started]
[10:32:07] [Compiler] [Error occurred while compiling]
[10:32:07] [Compile Errors] [
/private/var/folders/hn/9w3xn84x7t96r43k96hjtvyw0000gn/T/CP Editor-wLdlHX/sol.cpp:1:10: fatal error: 'bits/stdc++.h' file not found
#include <bits/stdc++.h>
         ^~~~~~~~~~~~~~~
1 error generated.
]`

But when I run the same file on command line with `g++ main.cpp -o main && ./main`, it run perfectly.

Can anyone help me pls ?

P/s : If anything, forgive my poor English, I'm learning though ...

**Comment by ouuan** [comment]
> [10:32:07] [Compiler] [Failed to open [/Volumes/External-Drive/Documents/C++ Projects/Proj3/script.cpp]. Do I have write permission?]

Does this error occur when you are not including `<bits/stdc++.h>`?

**Reply by Arius1408** [reply]
No sir

**Comment by Arius1408** [comment]
I don't think we have to focus on the 'permission' line, because I tried again with my cpp file on ~/Documents folder and it still responded the same way, just without the 'permission' line 

`[10:44:28] [Compiler] [Compilation has started]
[10:44:28] [Compiler] [Error occurred while compiling]
[10:44:28] [Compile Errors] [
/private/var/folders/hn/9w3xn84x7t96r43k96hjtvyw0000gn/T/CP Editor-Kkprid/sol.cpp:1:10: fatal error: 'bits/stdc++.h' file not found
#include <bits/stdc++.h>
         ^~~~~~~~~~~~~~~
1 error generated.
]`

**Comment by coder3101** [comment]
Can you show the error you get when setting compile command as `g++ -v` instead of `g++ -Wall`?

**Reply by Arius1408** [reply]
Bruh, sir ...

`[11:02:10] [Compiler] [Compilation has started]
[11:02:10] [Compiler] [Error occurred while compiling]
[11:02:10] [Compile Errors] [
Apple clang version 12.0.0 (clang-1200.0.32.29)
Target: x86_64-apple-darwin20.3.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
ignoring nonexistent directory "/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1"
 "/Library/Developer/CommandLineTools/usr/bin/clang" -cc1 -triple x86_64-apple-macosx11.0.0 -Wdeprecated-objc-isa-usage -Werror=deprecated-objc-isa-usage -Werror=implicit-function-declaration -emit-obj -mrelax-all -disable-free -disable-llvm-verifier -discard-value-names -main-file-name sol.cpp -mrelocation-model pic -pic-level 2 -mthread-model posix -mframe-pointer=all -fno-strict-return -masm-verbose -munwind-tables -target-sdk-version=11.1 -target-cpu penryn -dwarf-column-info -debugger-tuning=lldb -target-linker-version 609.8 -v -resource-dir /Library/Developer/CommandLineTools/usr/lib/clang/12.0.0 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -I /Users/airinishimiya/Documents -I/usr/local/include -stdlib=libc++ -internal-isystem /Library/Developer/CommandLineTools/usr/bin/../include/c++/v1 -internal-isystem /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/local/include -internal-isystem /Library/Developer/CommandLineTools/usr/lib/clang/12.0.0/include -internal-externc-isystem /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include -internal-externc-isystem /Library/Developer/CommandLineTools/usr/include -Wno-reorder-init-list -Wno-implicit-int-float-conversion -Wno-c99-designator -Wno-final-dtor-non-final-class -Wno-extra-semi-stmt -Wno-misleading-indentation -Wno-quoted-include-in-framework-header -Wno-implicit-fallthrough -Wno-enum-enum-conversion -Wno-enum-float-conversion -fdeprecated-macro -fdebug-compilation-dir /Users/airinishimiya/Documents -ferror-limit 19 -fmessage-length 0 -stack-protector 1 -fstack-check -mdarwin-stkchk-strong-link -fblocks -fencode-extended-block-signature -fregister-global-dtors-with-atexit -fgnuc-version=4.2.1 -fobjc-runtime=macosx-11.0.0 -fcxx-exceptions -fexceptions -fmax-type-align=16 -fdiagnostics-show-option -o /var/folders/hn/9w3xn84x7t96r43k96hjtvyw0000gn/T/sol-6d4206.o -x c++ "/private/var/folders/hn/9w3xn84x7t96r43k96hjtvyw0000gn/T/CP Editor-Kkprid/sol.cpp"
clang -cc1 version 12.0.0 (clang-1200.0.32.29) default target x86_64-apple-darwin20.3.0
ignoring nonexistent directory "/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/local/include"
ignoring nonexistent directory "/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/Library/Frameworks"
#include "..." search starts here:
#include <...> search starts here:
 /Users/airinishimiya/Documents
 /usr/local/include
 /Library/Developer/CommandLineTools/usr/bin/../include/c++/v1
 /Library/Developer/CommandLineTools/usr/lib/clang/12.0.0/include
 /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include
 /Library/Developer/CommandLineTools/usr/include
 /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks (framework directory)
End of search list.
/private/var/folders/hn/9w3xn84x7t96r43k96hjtvyw0000gn/T/CP Editor-Kkprid/sol.cpp:1:10: fatal error: 'bits/stdc++.h' file not found
#include <bits/stdc++.h>
         ^~~~~~~~~~~~~~~
1 error generated.
]`

**Reply by Arius1408** [reply]
I think I should go with the `g++ -Wall` for less error ...

**Reply by Arius1408** [reply]
[This](https://drive.google.com/file/d/1HoWf8G0n2UytPZZg9eCPEykWUSNX0RQV/view?usp=sharing) is an image of that act

**Reply by coder3101** [reply]
> I think I should go with the `g++ -Wall` for less error ...

No, the command was meant to put some more information like places where compiler looked for headers and the compiler in use.

**Reply by Arius1408** [reply]
Ok, so what should I do now ??
If you are going to tell me to install Xcode  and bits/stdc++.h file, I afraid that won't do btw, because I already did it ..

**Reply by coder3101** [reply]
The problem is that for some reason CP Editor is using Apple Clang compiler instead of `g++`. Maybe g++ is symlink-ed to Apple Clang.

I suggest you to use complete path to g++ like `/usr/local/bin/g++`, you can get the complete path of g++ by running `which g++`

**Reply by Arius1408** [reply]
So I should add the path to the variable environment ?

**Reply by coder3101** [reply]
> So I should add the path to the variable environment ?

Run from terminal `which g++` you will get where g++ is located. Change cpeditor compile command from `g++ -Wall` to `<output_you_got_from_which_command> -Wall`

**Reply by Arius1408** [reply]
Nice sir, thank u for your help !!!

**Comment by coder3101** [comment, marked as answer]
Actual Answer:

> The problem is that for some reason CP Editor is using Apple Clang compiler instead of `g++`. Maybe g++ is symlink-ed to Apple Clang.
>
> I suggest you to use complete path to g++ like `/usr/local/bin/g++`, you can get the complete path of g++ by running `which g++`
